### ISSUE TITLE
Change outbox record indexes and colums

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.NHibernate#144]
**This needs additional info** WIP

We have the following indexes:

1. PK, clustered
2. MessageId, non clustered
3. Dispatched, non clustered
4. DispatchedAt, non clustered

We have the following queries:

```
SELECT *
FROM OutBoxRecord
WHERE MessageId IN (x,y)
```
Uses: (2)

```
UPDATE {0}
SET Dispatched = true, DispatchedAt = :date
WHERE MessageId IN ( :messageid, :qualifiedMessageId ) And Dispatched = false
```
Uses (2) and (3)

```
DELETE from {0}
WHERE Dispatched = true And DispatchedAt < :date"
```

Uses (3) and (4)

This is non optimal as multiple indexes need to be queried when updating and deleting which results in unneeded locking and can result in occasional deadlocks.

## Proposal

- As Dispatched and DispatchedAt are either false and null or true and not null we can combine these in a single column resulting in one less index basically combining (3) and (4).
- The MessageId index maybe potentially improved if it included DispatchedAt value but as we are using NH this is not possible. The alternative is to add an additional index covering MessageId & DispatchedAt

## References

https://github.com/Particular/NServiceBus.NHibernate/blob/master/src/NServiceBus.NHibernate/Outbox/EntityMappings.cs

Generated SQL:

```sql
CREATE TABLE [dbo].[OutboxRecord] (
    [Id]                  BIGINT         IDENTITY (1, 1) NOT NULL,
    [MessageId]           NVARCHAR (255) NOT NULL,
    [Dispatched]          BIT            DEFAULT ((0)) NOT NULL,
    [DispatchedAt]        DATETIME       NULL,
    [TransportOperations] NVARCHAR (MAX) NULL,
    PRIMARY KEY CLUSTERED ([Id] ASC),
    UNIQUE NONCLUSTERED ([MessageId] ASC)
);


GO
CREATE NONCLUSTERED INDEX [OutboxRecord_Dispatched_Idx]
    ON [dbo].[OutboxRecord]([Dispatched] ASC);


GO
CREATE NONCLUSTERED INDEX [OutboxRecord_DispatchedAt_Idx]
    ON [dbo].[OutboxRecord]([DispatchedAt] ASC);


```

**Comment by ramonsmits** [comment]
Additional changes for consideration: https://groups.google.com/forum/#!topic/particularsoftware/CaZMSSti6BE

## TL;DR

- Use partitioning on DispatchedAt
- Improve cleanup logic

### Some background

> Our OutboxRecord table is stored in SQL Azure.  We have a high volume of traffic and we are seeing a large amount of database contention with selects, updates, and deletes.  The default installation of OutboxRecord is a table backed by a single database partition.  We're running a separate automation job to handle cleanup of the OutboxRecord table because the default delete logic is nearly always terminated by SQL Azure to avoid deadlocks of the other threads.  We have to delete from the table in smaller batches to avoid SQL terminating the process.

> A possible solution for this is to partition the OutboxRecord table on the DispatchedAt field.  We want to retain dispatched records for a period of time after which we can delete them.  This allows us to have a number of partitions, based on DispatchedAt, which allows us to drop entire partitions once we no longer need that data.  This avoids locks, but does change the design of OutboxRecord.  I have a working (in test) solution, but I had to modify the OutboxRecord table's indexes to make this work.

### Changes I had to make

> OutboxRecord has a clustered index on the ID field, but this field doesn't appear to be used by any queries.  MessageId is used extensively, so I dropped all indexes, created my partition function, and then created a new clustered index on MessageId *and* DispatchedAt.  I also created unclustered indexes on Dispatched and DispatchedAt.  I did not create an index for the ID field as it didn't seem to be used.

### And the question

> Are there any issues we may encounter with not having an index on ID (not to be confused with MessageId which is not part of a clustered index whereas before it was an unclustered index)?  I didn't see where ID was used, so I think I'm safe in removing the clustered index on that field, but wanted to check.

**Comment by SzymonPobiega** [comment]
@ramonsmits is the partitioning something we can implement or should we create a guidance around it?

**Comment by ramonsmits** [comment]
I checked if this issue is still relevant to see if the indexes have changed but the indexes are still the same.

**Comment by SzymonPobiega** [comment]
@ramonsmits frankly I have no idea how to proceed with it according to our processes. I suggest you raise an issue in PlatDev. I doubt it could be treated as maintainer budget.
